File: framework/languages/node/samples/TicTacToe.Ts/Server/Play/Infrastructure/ZLink/Actors/play-actor-transfer-adapter.ts
import { ZLinkMessage } from '@zlink-systems/framework';
import type { ZLinkActorTransferAdapter } from '@zlink-systems/framework';
import { PlayActor } from './play-actor';

interface PlayActorTransferState {
  readonly displayName: string;
  readonly level: number;
  readonly wins: number;
  readonly roomId?: string;
}

class PlayActorTransferAdapter implements ZLinkActorTransferAdapter<PlayActor> {
  async transferOut(actor: PlayActor): Promise<ZLinkMessage> {
    return ZLinkMessage.from<PlayActorTransferState>({
      displayName: actor.displayName,
      level: actor.level,
      wins: actor.wins,
      roomId: actor.roomId
    });
  }

  async transferIn(actorId: string, state: ZLinkMessage): Promise<PlayActor> {
    const restored = state.decode<PlayActorTransferState>();
    const actor = new PlayActor(actorId, restored.displayName, undefined, restored.level, restored.wins);
    actor.roomId = restored.roomId;
    return actor;
  }
}

export { PlayActorTransferAdapter };

File: framework/languages/dotnet/samples/TicTacToe/Server/Play/Infrastructure/ZLink/Actors/PlayActorTransferAdapter.cs
using TicTacToe.Shared.Contracts;
using Zlink.Framework.Contracts.Actors;
using Zlink.Framework.Contracts.Messaging;
using Zlink.Framework.Contracts.Spots;

namespace TicTacToe.Server.Play.Infrastructure.ZLink.Actors;

internal sealed class PlayActorTransferAdapter : IZLinkActorTransferAdapter<PlayActor>
{
    public ValueTask<ZLinkMessage> TransferOutAsync(
        PlayActor actor,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        return ValueTask.FromResult(ZLinkMessage.From(new PlayActorTransferState(
            actor.RoomId,
            actor.Player,
            actor.DestroyAfterEntrySpotJoin,
            actor.Disconnected)));
    }

    public ValueTask<PlayActor> TransferInAsync(
        string actorId,
        IZLinkActorContext context,
        ZLinkMessage state,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        var transferred = state.Decode<PlayActorTransferState>();
        var actor = new PlayActor(actorId, context);
        if (transferred.Player is not null) actor.ApplyPlayer(transferred.Player);
        if (!string.IsNullOrEmpty(transferred.RoomId)) actor.JoinRoom(transferred.RoomId);
        if (transferred.DestroyAfterEntrySpotJoin) actor.MarkForDestroyAfterRoomLeave();
        if (transferred.Disconnected) actor.MarkDisconnected();
        return ValueTask.FromResult(actor);
    }

    private sealed record PlayActorTransferState(
        string RoomId,
        PlayerInfo? Player,
        bool DestroyAfterEntrySpotJoin,
        bool Disconnected);
}
